changelog/8-1-2-Release.md

```diff
@@ -0,0 +1,1109 @@
+                        J.MethodInvocation m = super.visitMethodInvocation(method, ctx);
+                        // do something
+                        return m;
+                    }
+                });
+    }
+}
+```
+{% endtab %}
+{% endtabs %}
+
+### New approach to recipe visiting
+
+In the previous version of Rewrite, developers could use [Recipe.visit(List<SourceFile> ...)](https://github.com/openrewrite/rewrite/blob/v7.40.8/rewrite-core/src/main/java/org/openrewrite/Recipe.java#L477-L479) to randomly access different files when their recipes needed them. Unfortunately, this random access took a considerable amount of time and was not scalable for large source sets. To address that problem, Rewrite 8 introduces a new type of recipe that separates the run functionality into three, scalable, phases.
+
+For recipes that only change a single source file, you won't need to use the new recipe type. Instead, you will need to update the `Recipe.getVisitor()` method to be `public` instead of `protected`.
+
+For recipes that change many source files, you will need to convert the recipe to be a [ScanningRecipe](https://github.com/openrewrite/rewrite/blob/v8.1.1/rewrite-core/src/main/java/org/openrewrite/ScanningRecipe.java).
+
+#### What is a ScanningRecipe?
+
+A `ScanningRecipe` extends the normal `Recipe` and adds two key objects: an [accumulator](https://github.com/openrewrite/rewrite/blob/v8.1.1/rewrite-core/src/main/java/org/openrewrite/ScanningRecipe.java#L88-L90) and a [scanner](https://github.com/openrewrite/rewrite/blob/v8.1.1/rewrite-core/src/main/java/org/openrewrite/ScanningRecipe.java#L53). The `accumulator` object is a custom data structure defined by the recipe itself to store any information the recipe needs to function. The `scanner` object is a `visitor` which populates the `accumulator` with data.
+
+Scanning recipes have three phases:
+
+1. A scanning phase that collects information while making no new code changes. In this phase, the `scanner` is called for each source file and information is added to the `accumulator` that the recipe will need for future steps.
+    * For example, a recipe might want to detect whether a project is a Maven project or not. The `scanner` could detect a `pom.xml` file and add a flag to the `accumulator` so that future steps know this.
+2. An _optional_ generating phase where new files are created (if any are needed). In this phase, the `accumulator` can be accessed to determine whether or not a file should be created.
+3. An editing phase where the recipe makes changes (as it would before). Like the generating phase, an `accumulator` can be accessed to make changes – but you **can not** randomly access other source code or files.
+
+TODO: We need an example for this. Right now, the migration recipe links to nothing: https://github.com/openrewrite/rewrite/blob/main/rewrite-java-test/src/test/java/org/openrewrite/java/recipes/MigrateRecipeToRewrite8Test.java#L542 
```

Comment:
To help make reviewing easier - pointing out the TODOs that I feel like still need work.

Reply:
Thanks - I'll update this to include that example.